File: Problems/Heaps/Design-Twitter/Design-Twitter.cpp
#include <iostream>
#include <unordered_map>
#include <set>
#include <vector>
#include <queue>

// Using a heap - Time: O(n log n)

class Twitter {
public:
    std::unordered_map<int, std::set<int>> following;
    std::unordered_map<int, std::vector<std::pair<int, int>>> tweets;
    long long time;

    Twitter()
    {
        time = 0;
    }

    void postTweet(int userId, int tweetId)
    {
        tweets[userId].push_back({time++, tweetId});
    }

    std::vector<int> getNewsFeed(int userId)
    {
        std::vector<int> result;
        std::priority_queue<std::pair<int, int>> pq;

        // Push user's tweets into priority_queue
        for (auto t : tweets[userId])
        {
            pq.push(t);
        }

        // Push followed users' tweets into priority_queue
        for (auto f : following[userId])
        {
            for (auto t : tweets[f])
            {
                pq.push(t);
            }
        }

        // Retrieve top 10 tweets from priority_queue
        while (!pq.empty() && result.size() < 10)
        {
            result.push_back(pq.top().second);
            pq.pop();
        }

        return result;
    }

    void follow(int followerId, int followeeId)
    {
        following[followerId].insert(followeeId);
    }

    void unfollow(int followerId, int followeeId)
    {
        following[followerId].erase(followeeId);
    }
};

int main() {
    Twitter twitter;

    // Input
    std::cout << "Input: [\"Twitter\", \"postTweet\", \"getNewsFeed\", \"follow\", \"postTweet\", \"getNewsFeed\", \"unfollow\", \"getNewsFeed\"]" << std::endl;
    std::cout << "[[], [1, 5], [1], [1, 2], [2, 6], [1], [1, 2], [1]]" << std::endl;

    // Performing actions with actual data
    twitter.postTweet(1, 5);
    std::vector<int> feed1 = twitter.getNewsFeed(1);
    std::cout << "Output: [[";

    for (size_t i = 0; i < feed1.size(); ++i) {
        std::cout << feed1[i];
        if (i != feed1.size() - 1) {
            std::cout << ", ";
        }
    }

    std::cout << "], ";

    twitter.follow(1, 2);
    twitter.postTweet(2, 6);
    std::vector<int> feed2 = twitter.getNewsFeed(1);
    std::cout << "[";

    for (size_t i = 0; i < feed2.size(); ++i) {
        std::cout << feed2[i];
        if (i != feed2.size() - 1) {
            std::cout << ", ";
        }
    }

    std::cout << "], ";

    twitter.unfollow(1, 2);
    std::vector<int> feed3 = twitter.getNewsFeed(1);
    std::cout << "[";

    for (size_t i = 0; i < feed3.size(); ++i) {
        std::cout << feed3[i];
        if (i != feed3.size() - 1) {
            std::cout << ", ";
        }
    }

    std::cout << "]]" << std::endl;

    return 0;
}

File: Problems/Heaps/Design-Twitter/Design-Twitter.cs
using System;
using System.Collections.Generic;

// Using a heap - Time: O(n log n)

public class Twitter
{
    private Dictionary<int, HashSet<int>> following;
    private Dictionary<int, List<Tuple<int, int>>> tweets;
    private long time;

    public Twitter()
    {
        following = new Dictionary<int, HashSet<int>>();
        tweets = new Dictionary<int, List<Tuple<int, int>>>();
        time = 0;
    }

    public void PostTweet(int userId, int tweetId)
    {
        if (!tweets.ContainsKey(userId))
        {
            tweets[userId] = new List<Tuple<int, int>>();
        }
        tweets[userId].Add(new Tuple<int, int>((int)time++, tweetId));
    }

    public IList<int> GetNewsFeed(int userId)
    {
        var result = new List<int>();
        var comparer = new TupleComparer();
        var pq = new SortedSet<Tuple<int, int>>(comparer);

        if (tweets.ContainsKey(userId))
        {
            foreach (var t in tweets[userId])
            {
                pq.Add(t);
            }
        }

        if (following.ContainsKey(userId))
        {
            foreach (var f in following[userId])
            {
                if (tweets.ContainsKey(f))
                {
                    foreach (var t in tweets[f])
                    {
                        pq.Add(t);
                    }
                }
            }
        }

        foreach (var item in pq)
        {
            result.Add(item.Item2);
            if (result.Count == 10)
            {
                break;
            }
        }

        return result;
    }

    public void Follow(int followerId, int followeeId)
    {
        if (!following.ContainsKey(followerId))
        {
            following[followerId] = new HashSet<int>();
        }
        following[followerId].Add(followeeId);
    }

    public void Unfollow(int followerId, int followeeId)
    {
        if (following.ContainsKey(followerId))
        {
            following[followerId].Remove(followeeId);
        }
    }
}

public class TupleComparer : IComparer<Tuple<int, int>>
{
    public int Compare(Tuple<int, int> x, Tuple<int, int> y)
    {
        return y.Item1.CompareTo(x.Item1);
    }
}

class Program
{
    static void Main()
    {
        Twitter twitter = new Twitter();

        // Input
        Console.WriteLine("Input: [\"Twitter\", \"postTweet\", \"getNewsFeed\", \"follow\", \"postTweet\", \"getNewsFeed\", \"unfollow\", \"getNewsFeed\"]");
        Console.WriteLine("[[], [1, 5], [1], [1, 2], [2, 6], [1], [1, 2], [1]]");

        Console.Write("Output: [");

        

        // Performing actions with actual data
        twitter.PostTweet(1, 5);
        IList<int> feed1 = twitter.GetNewsFeed(1);

        foreach (var item in feed1)
        {
            Console.Write(item + " ");
        }

        // Console.Write("[" + string.Join(", ", feed1) + "]");

        twitter.Follow(1, 2);
        twitter.PostTweet(2, 6);
        IList<int> feed2 = twitter.GetNewsFeed(1);
        Console.Write("[" + string.Join(", ", feed2) + "]");

        twitter.Unfollow(1, 2);
        IList<int> feed3 = twitter.GetNewsFeed(1);
        Console.Write("[" + string.Join(",", feed3) + "]");

        Console.WriteLine("]");
    }
}

